Add mapbox zoom 1793

DrAcula27 opened the pull request:
Fixes #1793 

Creating PR per @ryanfchase request via Slack. Will likely change how popup is implemented.

  - [x] Up to date with `main` branch
  - [x] Branch name follows [guidelines](https://github.com/hackforla/311-data/blob/master/GETTING_STARTED.md#feature-branching)
  - [ ] All PR Status checks are successful
  - [ ] Peer reviewed and approved

Any questions? See the [getting started guide](https://github.com/hackforla/311-data/blob/master/GETTING_STARTED.md)

ryanfchase reviewed (changes requested):
Leaving further information here: https://github.com/hackforla/311-data/issues/1793#issuecomment-2418636581

DrAcula27 commented:
@ryanfchase @traycn I have updated this PR. The code is modified as follows:
- no longer using Mapbox.popup; instead using MUI Tooltip
- conditionally rendering MUI tooltip with the required text per design team OR the `Zoom out` title based on the state of the map.
  - if the `filterGeo` state is active, meaning a NC is selected, the MUI tooltip renders
  - if the `filterGeo` state is inactive, meaning no NC is selected, the `Zoom out` title for the zoom out button renders
- moved the tooltip to its own component, `ZoomTooltip` for easier code management, especially when we add styling 
  - https://mui.com/material-ui/react-tooltip/#customization
- added comments to clarify steps
- removed unnecessary comments around where I was working to keep code clean

ryanfchase commented:
Thanks @DrAcula27! I'll have a look at this soon
- ETA: 10/28/24 @ ~7pm PT

ryanfchase commented on components/Map/Map.jsx:
```diff
@@ -229,7 +287,14 @@ class Map extends React.Component {
+                filterGeo: geo,
+                minZoom: this.map.getZoom(),
+              });
+              this.map.setMinZoom(this.state.minZoom);
```
careful, you are not guaranteed to have the latest version of `this.state.minZoom` here. Either:

```js
this.setState({
  filterGeo: geo,
  minZoom: this.map.getZoom()
});
this.map.setMinZoom(this.map.getZoom())
```

or,
```js
this.setState((prevState) => {
  const newMinZoom = this.map.getZoom();
  this.map.setMinZoom(newMinZoom);
  return {
    filterGeo: geo,
    minZoom: newMinZoom,
})
```

I think I prefer the second option, since it guarantees you are setting minZoom to the map's zoom level *exactly at the time of setting the state*. But either way works, as long as you test your work and it still works as expected.


[source](https://medium.com/@minduladilthushan/resolving-the-state-not-updating-immediately-issue-in-react-js-febf5959c0cf#:~:text=When%20you%20update%20the%20state,return%20the%20old%20state%20value.)

DrAcula27 replied:
I think I agree with the second option as well. I will implement this one!

ryanfchase commented on components/Map/Map.jsx:
```diff
@@ -204,6 +222,45 @@ class Map extends React.Component {
+
+      // show the zoomtooltip on hover if the map is locked onto an ncLayer
+      const handleMouseEnter = () => {
+        if (this.state.filterGeo) {
```
could we additionally check to see if we are at the `minZoom` level as an additional check? My hunch is that design will prefer we only render the tooltip if we are hovering over zoom out _and_ we are at max zoom, but I could be convinced either way.

DrAcula27 replied:
From what I understand, `filterGeo` should only be truthy if we are at the `minZoom` level. With the change you mention before this one, I don't think changing this will be necessary.

ryanfchase replied:
I could be wrong, but I think `filterGeo` is insufficient for checking `minZoom`. To demonstrate, I've performed the following steps:
1. select an NC
2.  manually zoom in a little (I use scroll wheel here)
3.  hover over zoom out button
4. confirm that tooltip appears, despite us _not_ being at minZoom

<details><summary>GIF: reproducing the behavior above:</summary>
<p>

![zoom-out-hover-info](https://github.com/user-attachments/assets/27b6d0a6-79d9-4e86-a507-e18d3e35495a)


</p>
</details>

DrAcula27 replied:
@ryanfchase @traycn I misunderstood what behavior was desired here. I have fixed this issue!

### Now, the behavior is:

1. select a NC
    1. map auto-zooms onto it
    1. zoom out control shows disabled
1. hover over zoom out
    1. mouse cursor is `not-allowed`
    1. custom tooltip appears 
1. zoom in extra (I also used scroll wheel)
1. hover over zoom out again
    1. mouse cursor remains normal
    1. custom tooltip does not appear
1. zoom out back to current minZoom level with the NC selected
    1. control goes disabled again
1. hover over zoom out again
    1. custom tooltip appears

ryanfchase commented on components/Map/zoomTooltip.jsx:
```diff
@@ -0,0 +1,39 @@
+    >
+      {/* empty span for positioning the zoomtooltip */}
+      <span
+        className="mapboxgl-ctrl-icon minus-sign-clone"
```
So smart! Nothing to do here, I just like this over my implementation. You can mark this as resolved 😄

ryanfchase reviewed (changes requested):
I think we need to undo the code styling. A small amount is OK, but there are good chunks of code added to this review that are unrelated -- I want to be sure I'm not distracted by code styling changes. We should do a Prettier integration as an upcoming code health feature, and then you can sync your style settings with our style guide.

Otherwise, this is great work!

DrAcula27 commented:
@ryanfchase I am not sure how to remove the crazy amount of changes made as part of this PR due to the Vite migration. I will research this further, but any help is appreciated!

DrAcula27 commented:
Resolved extra commits problem. Closing this PR in favor of https://github.com/hackforla/311-data/pull/1860
